runtime/onert/api/src/nnfw_api_internal.cc

```diff
@@ -492,27 +492,26 @@ NNFW_STATUS nnfw_session::apply_tensorinfo(uint32_t index, nnfw_tensorinfo ti)
     }
   }
 
-  auto ind = primary_subgraph()->getInputs().at(index);
-  auto &input = primary_subgraph()->operands().at(ind);
-
-  onert::ir::Shape new_shape(ti.rank);
-  for (int32_t i = 0; i < ti.rank; i++)
-    new_shape.dim(i) = ti.dims[i];
-
-  // if passed shape is same with the shape of model, do nothing
-  if (input.info().shape() == new_shape)
-    return NNFW_STATUS_NO_ERROR;
-
   if (!isStatePreparedOrFinishedRun())
   {
     // In this case, if we apply input shape in primary_subgraph, it will propagate after
     // compilation and excution
+    auto ind = primary_subgraph()->getInputs().at(index);
+    auto &input = primary_subgraph()->operands().at(ind);
+
+    onert::ir::Shape new_shape(ti.rank);
+    for (int32_t i = 0; i < ti.rank; i++)
+      new_shape.dim(i) = ti.dims[i];
 
     // overwrite input shape with the shape from ti
     input.info().shape(new_shape);
   }
   else // when called after nnfw_session::prepare()
   {
+    onert::ir::Shape new_shape(ti.rank);
+    for (int32_t i = 0; i < ti.rank; i++)
+      new_shape.dim(i) = ti.dims[i];
+
     _execution->changeInputShape(onert::ir::IOIndex(index), new_shape);
```

Comment:
Adding a check `if (_execution->getInputShape(onert::ir::IOIndex{index}) != new_shape)` will be good optimization that prevents activation of dynamic mode when not necessary. Unless it is proved to break something

Reply:
Thanks. Fixed.

Reply:
I have some concern with this change. Could you please refer to https://github.com/Samsung/ONE/pull/4718#issuecomment-714894837?

One other thing is that what `_execution->changeInputShape(..)`  does is simply entering new shape information into a map inside `IODescription`. IMHO elimination of calling  `_execution->changeInputShape(..)` by adding one more condition (`if (_execution->getInputShape(onert::ir::IOIndex{index}) != new_shape)`) does not provide optimization.

Reply:
The main purpose of this optimization is to **prevent creating** an item in that map when two conditions are met: a) the tensor is not yet registered in the map of dynamic tensors, b) the new shape is the same as the static shape. Of course, there are many ways to implement that logic, and moving it from nnfw_api to the inside of `changeInputShape` would be even more appropriate. But following the idea of PR #3998, moving the comparison from the wrong place to the right place would fix the bug with minimum code change.

Reply:
The purpose of this PR is not to follow the optimization of #3998 but to follow the rule 1) and 2) in https://github.com/Samsung/ONE/issues/4625#issue-721225511.

What you propose is one way to solve current crash but I mentioned in
https://github.com/Samsung/ONE/pull/4718#issuecomment-714894837, it is difficult to communicate. 

I tried to write down how I will communicate if I accept your suggestion and it is complex:
- If input shape is changed, ONERT treats the input tensor as dynamic
- But once the input tensor becomes dynamic, the input will remain dynamic if the shape changes. (see 5th run in https://github.com/Samsung/ONE/pull/4718#issuecomment-714894837)

Also, IMHO another drawback with your suggestion is that... when user creates a model with Tensorflow with shape e.g., `(None, 100)`, converted tflite file will have `(1, 100)` shape. However, users won't know the `None` will be converted to to `1` in tflite file. So, they will have hard time to understand why the shape was not changed with 2nd run and it is static... and why 4, 5th run is dynamic. It is too difficult to understand.

So, IMHO, I think rules in https://github.com/Samsung/ONE/issues/4625#issue-721225511 and the change in this PR is better for API.